File: java/src/main/java/com/alxarafe/app/domain/inventory/valueobject/Quantity.java
package com.alxarafe.app.domain.inventory.valueobject;

import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Quantity stored internally as a {@code long} scaled by 10^-SCALE to avoid
 * floating point drift (decisión M5).
 *
 * <p>Floating point is only used on the HTTP boundary
 * ({@link #fromDecimal(double, String)}), which rounds HALF_UP to the scale.
 * Arithmetic and database reads always work with exact integers
 * ({@link #fromScaled(long, String)} / {@link #fromDecimalString(String, String)}).
 *
 * <p>The unit is not part of the arithmetic: any unit is accepted, and two
 * quantities combined must share the same unit.
 */
public final class Quantity {

    public static final int SCALE = 6;
    private static final long SCALE_FACTOR = 1_000_000L;
    private static final Pattern DECIMAL = Pattern.compile("^(\\d+)(?:\\.(\\d+))?$");

    private final long scaledUnits;
    private final String unit;

    private Quantity(long scaledUnits, String unit) {
        if (scaledUnits <= 0) {
            throw new IllegalArgumentException("Quantity must be strictly positive.");
        }
        Objects.requireNonNull(unit, "Quantity unit cannot be null.");
        if (unit.isBlank()) {
            throw new IllegalArgumentException("Quantity unit cannot be empty.");
        }
        this.scaledUnits = scaledUnits;
        this.unit = unit;
    }

    /**
     * Creates a quantity from a decimal value (HTTP boundary, HALF_UP rounding).
     */
    public static Quantity fromDecimal(double value, String unit) {
        if (!Double.isFinite(value)) {
            throw new IllegalArgumentException("Quantity must be a finite number.");
        }
        if (value <= 0.0) {
            throw new IllegalArgumentException("Quantity must be strictly positive. Got: " + value);
        }
        long scaled = Math.round(value * SCALE_FACTOR);
        if (scaled <= 0) {
            throw new IllegalArgumentException("Quantity must be strictly positive. Got: " + value);
        }
        return new Quantity(scaled, unit);
    }

    /**
     * Creates a quantity from an already scaled integer (exact, no floating point).
     */
    public static Quantity fromScaled(long scaledUnits, String unit) {
        return new Quantity(scaledUnits, unit);
    }

    /**
     * Creates a quantity from a PostgreSQL NUMERIC(18,6) string like '30.500000'.
     */
    public static Quantity fromDecimalString(String decimal, String unit) {
        Objects.requireNonNull(decimal, "Decimal quantity cannot be null.");
        var matcher = DECIMAL.matcher(decimal);
        if (!matcher.matches()) {
            throw new IllegalArgumentException("Invalid decimal quantity: " + decimal);
        }
        long whole = Long.parseLong(matcher.group(1));
        String fraction = matcher.group(2) == null ? "" : matcher.group(2);
        if (fraction.length() > SCALE) {
            fraction = fraction.substring(0, SCALE);
        }
        fraction = fraction + "0".repeat(SCALE - fraction.length());
        long scaled = Math.multiplyExact(whole, SCALE_FACTOR) + Long.parseLong(fraction);
        return new Quantity(scaled, unit);
    }

    public long scaledUnits() {
        return scaledUnits;
    }

    public String unit() {
        return unit;
    }

    public double value() {
        return (double) scaledUnits / (double) SCALE_FACTOR;
    }

    /**
     * Exact decimal representation in NUMERIC(15,6) form for DB writes.
     */
    public String toDecimalString() {
        return String.format("%d.%06d", scaledUnits / SCALE_FACTOR, scaledUnits % SCALE_FACTOR);
    }

    public Quantity add(Quantity other) {
        Objects.requireNonNull(other, "Quantity to add cannot be null.");
        assertSameUnit(other);
        return new Quantity(this.scaledUnits + other.scaledUnits, this.unit);
    }

    public Quantity subtract(Quantity other) {
        Objects.requireNonNull(other, "Quantity to subtract cannot be null.");
        assertSameUnit(other);
        if (this.scaledUnits < other.scaledUnits) {
            throw new IllegalArgumentException(
                    "Insufficient quantity for subtraction. Has: " + this.value() + ", wants: " + other.value());
        }
        return new Quantity(this.scaledUnits - other.scaledUnits, this.unit);
    }

    private void assertSameUnit(Quantity other) {
        if (!this.unit.equals(other.unit)) {
            throw new IllegalArgumentException(
                    "Cannot combine quantities with different units: " + this.unit + " vs " + other.unit);
        }
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof Quantity quantity)) {
            return false;
        }
        return scaledUnits == quantity.scaledUnits && unit.equals(quantity.unit);
    }

    @Override
    public int hashCode() {
        return Objects.hash(scaledUnits, unit);
    }

    @Override
    public String toString() {
        return value() + " " + unit;
    }
}

File: php/src/Domain/Inventory/ValueObject/Quantity.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Domain\Inventory\ValueObject;

use InvalidArgumentException;

/**
 * Quantity stored internally as an integer scaled by 10^-SCALE to avoid
 * floating point drift (decisión M5).
 *
 * Solo se usan números de coma flotante en la frontera HTTP
 * (Quantity::fromDecimal), que redondea HALF_UP a la escala definida.
 * La aritmética y las lecturas de base de datos trabajan siempre con
 * enteros exactos (fromScaled/fromDecimalString).
 *
 * La unidad no interviene en la aritmética: se admite cualquier unidad,
 * que debe coincidir entre las cantidades que se combinen.
 */
final readonly class Quantity
{
    public const int SCALE = 6;
    private const int SCALE_FACTOR = 1_000_000;

    private function __construct(
        private int $scaledUnits,
        private string $unit,
    ) {
        if ($scaledUnits <= 0) {
            throw new InvalidArgumentException(
                'Quantity must be strictly positive.'
            );
        }
        if ($unit === '') {
            throw new InvalidArgumentException('Quantity unit cannot be empty.');
        }
    }

    /**
     * Crea la cantidad desde un valor decimal (frontera JSON, redondeo HALF_UP).
     */
    public static function fromDecimal(int|float $value, string $unit): self
    {
        if (is_float($value) && !is_finite($value)) {
            throw new InvalidArgumentException('Quantity must be a finite number.');
        }
        if ($value <= 0) {
            throw new InvalidArgumentException(
                'Quantity must be strictly positive. Got: ' . $value
            );
        }
        $scaled = (int) round($value * self::SCALE_FACTOR, 0, PHP_ROUND_HALF_UP);
        return new self($scaled, $unit);
    }

    /**
     * Crea la cantidad desde entero ya escalado (exacto, sin coma flotante).
     */
    public static function fromScaled(int $scaledUnits, string $unit): self
    {
        return new self($scaledUnits, $unit);
    }

    /**
     * Crea la cantidad desde un NUMERIC(18,6) de PostgreSQL ('30.500000').
     */
    public static function fromDecimalString(string $decimal, string $unit): self
    {
        if (preg_match('/^(\d+)(?:\.(\d+))?$/', $decimal, $matches) !== 1) {
            throw new InvalidArgumentException('Invalid decimal quantity: ' . $decimal);
        }
        $whole = (int) $matches[1];
        $fraction = str_pad(substr($matches[2] ?? '', 0, self::SCALE), self::SCALE, '0');
        return new self($whole * self::SCALE_FACTOR + (int) $fraction, $unit);
    }

    public function scaledUnits(): int
    {
        return $this->scaledUnits;
    }

    public function unit(): string
    {
        return $this->unit;
    }

    public function value(): float
    {
        return $this->scaledUnits / self::SCALE_FACTOR;
    }

    /**
     * Representación decimal exacta tipo NUMERIC(15,6) para escritura en BD.
     */
    public function toDecimalString(): string
    {
        return sprintf(
            '%d.%06d',
            intdiv($this->scaledUnits, self::SCALE_FACTOR),
            $this->scaledUnits % self::SCALE_FACTOR
        );
    }

    public function equals(self $other): bool
    {
        return $this->scaledUnits === $other->scaledUnits
            && $this->unit === $other->unit;
    }

    public function add(self $other): self
    {
        $this->assertSameUnit($other);
        return new self($this->scaledUnits + $other->scaledUnits, $this->unit);
    }

    public function subtract(self $other): self
    {
        $this->assertSameUnit($other);
        if ($this->scaledUnits < $other->scaledUnits) {
            throw new InvalidArgumentException(
                "Insufficient quantity for subtraction. Has: {$this->value()}, wants: {$other->value()}"
            );
        }
        return new self($this->scaledUnits - $other->scaledUnits, $this->unit);
    }

    private function assertSameUnit(self $other): void
    {
        if ($this->unit !== $other->unit) {
            throw new InvalidArgumentException(
                "Cannot combine quantities with different units: {$this->unit} vs {$other->unit}"
            );
        }
    }
}
